### PR TITLE
always fail outgoing message for stopped contacts

### PR DESCRIPTION
**Pull request by norkans7**
Fix
1) https://github.com/rapidpro/rapidpro/issues/401
2) https://github.com/rapidpro/rapidpro/issues/391

also Add channel logs for these cases

### CODE REVIEW

**Comment by nicpottier** [comment]
Are we properly removing triggers for contacts that are stopped/blocked?

**Comment by norkans7** [comment]
the message will not be failed so it will not be sent

**Comment by nicpottier** [comment]
But the trigger will still exist, which is weird. And we'll keep generating failed messages no?

**Comment by norkans7** [comment]
now when a contact is blocked or stopped we clear the contact on the triggers that have it

File: temba/msgs/models.py [hunk 1]
```diff
@@ -1121,6 +1125,11 @@ def resend(self):
         if cloned.broadcast:
             cloned.broadcast.update()
 
+        if status == FAILED:
+            ChannelLog.log_error(dict_to_struct('MockMsg', cloned.as_task_json()),
+                                 "Can't send message to %s contact" % ("stopped" if cloned.contact.is_stopped
+                                                                       else "blocked"))
```
Comment (ericnewcomer) [inline temba/msgs/models.py]: There's other ways a message can be failed out side of stopped and blocked. Are you sure this is right?

File: temba/msgs/models.py [hunk 2]
```diff
@@ -1394,7 +1406,13 @@ def create_outgoing(cls, org, user, recipient, text, broadcast=None, channel=Non
         if topup_id is not None:
             msg_args['topup_id'] = topup_id
 
-        return Msg.objects.create(**msg_args) if insert_object else Msg(**msg_args)
+        msg = Msg.objects.create(**msg_args) if insert_object else Msg(**msg_args)
+
+        if status == FAILED:
+            ChannelLog.log_error(dict_to_struct('MockMsg', msg.as_task_json()),
+                                 "Can't send message to %s contact" % ("stopped" if contact.is_stopped else "blocked"))
```
Comment (ericnewcomer) [inline temba/msgs/models.py]: Same thing here?

File: temba/msgs/models.py (hunk 1 above)
Comment (norkans7) [inline temba/msgs/models.py]: this is to prevent resending a failed msg to a stopped/blocked contact; we fail the outgoing message directly with this

Comment (norkans7) [inline temba/msgs/models.py]: here the status was always set to PENDING since to force a resend

**Comment by norkans7** [comment]
The only way now to send to a blocked/stopped contact is by explicitly specifying its urn in the omnibox or click the specific urn on the contact read page

**Comment by norkans7** [comment]
I just left one way for the user who really want to send a message to click explicitly on the URN to send to

**Comment by norkans7** [comment]
This will behave the same way as if the user explicitly enter the URN from the inbox page send message modal or including the URN when sending to a different contact/group

File: temba/contacts/models.py [hunk 3]
```diff
@@ -1431,6 +1432,8 @@ def stop(self, user):
 
         self.clear_all_groups(get_anonymous_user())
 
+        self.trigger_set.clear()
```
Comment (nicpottier) [inline temba/contacts/models.py]: If there is a trigger on a single contact that is blocked, what happens to that trigger? Do we still show it in the UI? How?

Comment (norkans7) [inline temba/contacts/models.py]: update to archived the contact that were attached only to the blocked contact

File: temba/contacts/models.py [hunk 4]
```diff
@@ -1390,6 +1390,16 @@ def apply_action_unstop(cls, user, contacts):
             changed.append(contact.pk)
         return changed
 
+    def archive_contact_empty_triggers(self):
```
Comment (nicpottier) [inline temba/contacts/models.py]: This looks right but really should be a class method on Trigger instead (contacts don't depend on triggers, but triggers depend on contacts)

Can call it `archive_triggers_for_contact`

**Pull request merged**